Use of rounding fails with error evaluating field

The forms below passed validation with Validate 1.4.9, but they are failing with Validate 1.5.0 (with JavaRosa 2.2.0).

* [round.xlsx](https://github.com/opendatakit/javarosa/files/1078288/round.xlsx)
* [round.xml.txt](https://github.com/opendatakit/javarosa/files/1078287/round.xml.txt)

```
java.lang.RuntimeException: Error evaluating field 'calc': For input string: "���"
	at org.javarosa.core.model.LatestDagBase.evaluateTriggerable(LatestDagBase.java:139)
	at org.javarosa.core.model.LatestDagBase.doEvaluateTriggerables(LatestDagBase.java:63)
	at org.javarosa.core.model.Safe2014DagImpl.evaluateTriggerables(Safe2014DagImpl.java:348)
	at org.javarosa.core.model.Safe2014DagImpl.initializeTriggerables(Safe2014DagImpl.java:434)
	at org.javarosa.core.model.Safe2014DagImpl.initializeTriggerables(Safe2014DagImpl.java:412)
	at org.javarosa.core.model.FormDef.initializeTriggerables(FormDef.java:709)
	at org.javarosa.core.model.FormDef.initialize(FormDef.java:1251)
	at org.opendatakit.validate.FormValidator.validate(FormValidator.java:445)
	at org.opendatakit.validate.FormValidator.validate(FormValidator.java:343)
	at org.opendatakit.validate.FormValidator.actionPerformed(FormValidator.java:254)
	at javax.swing.AbstractButton.fireActionPerformed(AbstractButton.java:2022)
	at javax.swing.AbstractButton$Handler.actionPerformed(AbstractButton.java:2348)
	at javax.swing.DefaultButtonModel.fireActionPerformed(DefaultButtonModel.java:402)
	at javax.swing.DefaultButtonModel.setPressed(DefaultButtonModel.java:259)
	at javax.swing.plaf.basic.BasicButtonListener.mouseReleased(BasicButtonListener.java:252)
	at java.awt.Component.processMouseEvent(Component.java:6533)
	at javax.swing.JComponent.processMouseEvent(JComponent.java:3324)
	at java.awt.Component.processEvent(Component.java:6298)
	at java.awt.Container.processEvent(Container.java:2236)
	at java.awt.Component.dispatchEventImpl(Component.java:4889)
	at java.awt.Container.dispatchEventImpl(Container.java:2294)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.LightweightDispatcher.retargetMouseEvent(Container.java:4888)
	at java.awt.LightweightDispatcher.processMouseEvent(Container.java:4525)
	at java.awt.LightweightDispatcher.dispatchEvent(Container.java:4466)
	at java.awt.Container.dispatchEventImpl(Container.java:2280)
	at java.awt.Window.dispatchEventImpl(Window.java:2746)
	at java.awt.Component.dispatchEvent(Component.java:4711)
	at java.awt.EventQueue.dispatchEventImpl(EventQueue.java:758)
	at java.awt.EventQueue.access$500(EventQueue.java:97)
	at java.awt.EventQueue$3.run(EventQueue.java:709)
	at java.awt.EventQueue$3.run(EventQueue.java:703)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:90)
	at java.awt.EventQueue$4.run(EventQueue.java:731)
	at java.awt.EventQueue$4.run(EventQueue.java:729)
	at java.security.AccessController.doPrivileged(Native Method)
	at java.security.ProtectionDomain$JavaSecurityAccessImpl.doIntersectionPrivilege(ProtectionDomain.java:80)
	at java.awt.EventQueue.dispatchEvent(EventQueue.java:728)
	at java.awt.EventDispatchThread.pumpOneEventForFilters(EventDispatchThread.java:201)
	at java.awt.EventDispatchThread.pumpEventsForFilter(EventDispatchThread.java:116)
	at java.awt.EventDispatchThread.pumpEventsForHierarchy(EventDispatchThread.java:105)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:101)
	at java.awt.EventDispatchThread.pumpEvents(EventDispatchThread.java:93)
	at java.awt.EventDispatchThread.run(EventDispatchThread.java:82)
Caused by: java.lang.NumberFormatException: For input string: "���"
	at sun.misc.FloatingDecimal.readJavaFormatString(FloatingDecimal.java:2043)
	at sun.misc.FloatingDecimal.parseDouble(FloatingDecimal.java:110)
	at java.lang.Double.parseDouble(Double.java:538)
	at java.lang.Double.valueOf(Double.java:502)
	at org.javarosa.xpath.expr.XPathFuncExpr.round(XPathFuncExpr.java:1051)
	at org.javarosa.xpath.expr.XPathFuncExpr.eval(XPathFuncExpr.java:199)
	at org.javarosa.xpath.XPathConditional.evalRaw(XPathConditional.java:71)
	at org.javarosa.core.model.condition.Triggerable.evalRaw(Triggerable.java:219)
	at org.javarosa.core.model.condition.Recalculate.eval(Recalculate.java:56)
	at org.javarosa.core.model.condition.Triggerable.apply(Triggerable.java:152)
	at org.javarosa.core.model.LatestDagBase.evaluateTriggerable(LatestDagBase.java:125)
	... 45 more
```

--- a/src/org/javarosa/xpath/expr/XPathFuncExpr.java
+++ b/src/org/javarosa/xpath/expr/XPathFuncExpr.java
@@ -200,8 +200,7 @@ public class XPathFuncExpr extends XPathExpression {
                 assertArgsCount(name, args, 2);
                 places = toNumeric(argVals[1]).intValue();
             }
-            // Some locales use ',' instead of '.' as a decimal separator, but not supported by Decimal class.
-            return round(toNumeric(argVals[0].toString().replace(',','.')), places);
+            return round(toNumeric(argVals[0]), places);
         } else if (name.equals("string")) {
             assertArgsCount(name, args, 1);
             return toString(argVals[0]);
@@ -656,18 +655,18 @@ public class XPathFuncExpr extends XPathExpression {
              * when converting a string to a number
              */
 
-            String s = (String)o;
-            double d;
+            final String s = ((String) o)
+                    .replace(',', '.') // Some locales use ',' instead of '.'
+                    .trim();
+
             try {
-                s = s.trim();
                 for (int i = 0; i < s.length(); i++) {
                     char c = s.charAt(i);
                     if (c != '-' && c != '.' && (c < '0' || c > '9'))
                         throw new NumberFormatException();
                 }
 
-                d = Double.parseDouble(s);
-                val = d;
+                val = Double.parseDouble(s);
             } catch (NumberFormatException nfe) {
                 val = NaN;
             }
@@ -679,9 +678,9 @@ public class XPathFuncExpr extends XPathExpression {
 
         if (val != null) {
             return val;
-        } else {
-            throw new XPathTypeMismatchException("converting to numeric");
         }
+
+        throw new XPathTypeMismatchException("converting to numeric");
     }
 
     /**

--- a/test/org/javarosa/xpath/test/XPathEvalTest.java
+++ b/test/org/javarosa/xpath/test/XPathEvalTest.java
@@ -335,6 +335,8 @@ public class XPathEvalTest extends TestCase {
         testEval("round('12345.12345', -2)", 12300.0);
         testEval("round('12350.12345', -2)", 12400.0);
         testEval("round('12345.12345', -3)", 12000.0);
+        // round, with a comma instead of a decimal point
+        testEval("round('4,6'            )", 5.0);
         // XPath specification tests
         testEval("round('1 div 0', 0)", NaN);
         testEval("round('14.5')", 15.0);

